### ISSUE TITLE
PuppetJS and Partials

### ISSUE THREAD

**Issue by careri** [Starcounter/Starcounter#3844]
Starcounter: 2.2.1834

I'm using the partial page pattern from the KitchenSink example.

I click a link in my meny that shall replace the CurrentPage in the json-model. This works fine for most of my partials.

But I have one that behaves strange.

First I get a call to replace a path that will be rendered as a grid using hot-table:
![image](https://cloud.githubusercontent.com/assets/4263597/18634610/0000d744-7e82-11e6-98bb-4467856d2433.png)

This first call works just fine, then almost directly after I get a second call where the obj-variable is null.
![image](https://cloud.githubusercontent.com/assets/4263597/18634644/25079ae6-7e82-11e6-96d7-594d22b5a76a.png)

If I go up one step in the call stack the debugger shows a value for ```that.obj``` 
![image](https://cloud.githubusercontent.com/assets/4263597/18634830/fa1bf7fe-7e82-11e6-8de6-893bb1f79a82.png)

Hitting F8 to continue yields the following error:
puppet.js:445 Uncaught TypeError: Cannot read property 'Selectables' of undefined

If I hit F5 to reload the page to the url it works just fine.

So for some reason the .obj-property seems to be set to null for a short while, during this period a patch is received from the server. This patch fails because of the model is null.

This is as far as I get. Is it possible to monitor the value change of .obj property?

Is this something that @warpech is best suited to help out with?

**Comment by per-samuelsson** [comment]
>Is this something that @warpech is best suited to help out with?

Probably. Also pinging @miyconst.

**Comment by per-samuelsson** [comment]
>Also pinging @miyconst.

And @tomalec :sunglasses:

**Comment by careri** [comment]
I did another check, and it's not a second call to:
```c#
function markObjPropertyByPath(obj, path) {
    var keys = path.split('/');
    var len = keys.length;
    if (len > 2) {
      for (var i = 1; i < len - 1; i++) {
        obj = obj[keys[i]];
      }
    }
    recursiveMarkObjProperties(obj[keys[len - 1]], len > 1? obj : undefined);
  }
```
It's actually the for-loop that replaces the value of obj.
The problem seems to be that the value of CurrentPage\CurrentPage\HAD has not yet changed it's value to the new page. But it still holds the Json for the previous page.

So the patch recieved from the server tries to patch a model that has yet to be assigned to the client json.

**Comment by careri** [comment]
I've added some logging to the Puppet.js
And the patches recieved from the server looks fine. The error seems to be the patch that shall replace "HAD" on the current page.
This actually does nothing, so the next patch that expects HAD to have the new value fails.
The patches received. 
>
=====================Received:
[{"op":"replace","path":"/_ver#s","value":6},{"op":"test","path":"/_ver#c$","value":0},{"op":"replace","path":"/ContainerPage/CurrentPage","value":{"HAD":{"Html":"/HAD/Logs.html","IsWorkstationLogs$":false,"Workstations":{"Selectables":[],"SelectedIndex$":0},"Levels":{"Selectables":[],"SelectedIndex$":0},"Filter$":"","FromTime$":"2016-09-20 10:54:12","Items":[]},"Html":"/sc/htmlmerger?HAD=/HAD/Logs.html"}},{"op":"replace","path":"/ContainerPage/CurrentPage/HAD/Levels/Selectables","value":[]}]
==============================
>
puppet.js:656 	 Processing:
{"op":"replace","path":"/_ver#s","value":6}
****
>
puppet.js:656 	 Processing:
{"op":"test","path":"/_ver#c$","value":0}
****
>
puppet.js:656 	 Processing:
{"op":"replace","path":"/ContainerPage/CurrentPage","value":{"HAD":{"Html":"/HAD/Logs.html","IsWorkstationLogs$":false,"Workstations":{"Selectables":[],"SelectedIndex$":0},"Levels":{"Selectables":[],"SelectedIndex$":0},"Filter$":"","FromTime$":"2016-09-20 10:54:12","Items":[]},"Html":"/sc/htmlmerger?HAD=/HAD/Logs.html"}}
****
>
puppet.js:656 	 Processing:
{"op":"replace","path":"/ContainerPage/CurrentPage/HAD/Levels/Selectables","value":[]}
****

This is the failing patch:
> {"op":"replace","path":"/ContainerPage/CurrentPage/HAD/Levels/Selectables","value":[]}

And it fails because the previous patch didn't replace the value of HAD

**Comment by careri** [comment]
I figured out what goes wrong. The server seems to send the wrong revision number for the patch. So this method places my patch in the waiting array instead of applying it:

My patch ends up in ```// add sequence to queue in correct position.```

```js
JSONPatchQueue.prototype.receive = function(obj, versionedJsonPatch){
	var consecutivePatch = versionedJsonPatch.slice(0);
	// strip Versioned JSON Patch specyfiv operation objects from given sequence
		if(this.purist){
			var testRemote = consecutivePatch.shift();
		}
		var replaceRemote = consecutivePatch.shift(),
			newRemoteVersion = replaceRemote.value;

	// TODO: perform versionedPath validation if needed (tomalec)

	if( newRemoteVersion <= this.remoteVersion){
	// someone is trying to change something that was already updated
    	throw new Error("Given version was already applied.");
	} else if ( newRemoteVersion == this.remoteVersion + 1 ){
	// consecutive new version
		while( consecutivePatch ){// process consecutive patch(-es)
			this.remoteVersion++;
			this.apply(obj, consecutivePatch);
			consecutivePatch = this.waiting.shift();
		}
	} else {
	// add sequence to queue in correct position.
		this.waiting[newRemoteVersion - this.remoteVersion -2] = consecutivePatch;
	}
};
```

Then the crashing method assumes that the patch is applied, which isn't. BOOM

So this might actually be something for @chrhol or @alemoi

**Comment by warpech** [comment]
Just to make sure that we are not investigating an already fixed issue, could you please try the current nightly?

We did some changes to queue processing in the last few weeks.

If the problem persists, could you package the app and send it to me for investigation with a zipped database?

**Comment by careri** [comment]
My app works without any data. So you should be fine just starting from scratch.

I'm running 2.2.1834. What nightly are you refering to?

**Comment by warpech** [comment]
2.2.1.3240 from http://downloads.starcounter.com/download

We are planning to make a new stable release this week.

**Comment by careri** [comment]
Ok, didn't know if you were referring to 2.3 or 2.2.
I'll give it a spin

**Comment by careri** [comment]
#3696 Argh, my code no longer compiles...

**Comment by tomalec** [comment]
As @warpech said, I'd also suggest updating to latest version, as we have removed problematic `markObjPropertyByPath` entirely.

But to get the better understanding on what is going on with what you describe in you previous comment  https://github.com/Starcounter/Starcounter/issues/3844#issuecomment-248262316, could you provide a full list of patches being exchanged?

**Comment by warpech** [comment]
We are investigating on Hangouts. @tomalec I will let you know if the problem is reduced to patches. Probably just the symptom is in patches.

**Comment by careri** [comment]
I'm in contact with @warpech  right now. Sending him the code.
The client no longer crashes, but it populates the waiting array with all patches sent after the problematic one.

**Comment by warpech** [comment]
@chrhol can you pls take a look? This is definitely a server side issue. 

When clicking on "Backups" and then "Logs", I get the following `_ver#s`: 1, 4, 3, 3. 

There are at least 2 problematic symptoms as we can see on the screenshot:

 - version 2 is never sent
-  version 3 is pushed twice at the same timestamp, with different content

I guess @chrhol will need access to the app source code.

![2016-09-20 13_30_23-start page - microsoft visual studio](https://cloud.githubusercontent.com/assets/566463/18668599/a53cf0da-7f36-11e6-840f-c11a6ec136d8.png)

**Comment by careri** [comment]
I stepped through the server code and I can confirm that the same local version is used twice:
src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs
```cs
size += GetSizeOfIntAsUtf8(changeLog.Version.LocalVersion);
```

Guess it's a sync issue when executing on multiple threads?

**Comment by chrhol** [comment]
Alright, I have gotten the app from @careri and have reproduced it. Not sure what the problem is at the moment, but there is a lot of pushes taking place (Session.CalculatePatchAndPushOnWebSocket())

> Guess it's a sync issue when executing on multiple threads?

Yes, for some reason. It shouldn't be though, since only one thread should have access to the session at a time. Maybe we have missed some pattern that is bypassing the lock. I will investigate more when I have time. Will be occupied in meetings the whole day though so I'm not sure exactly when I will get the time to do it.

**Comment by chrhol** [comment]
I found the problem. It's not directly a sync-issue, but a problem of recursive calls to generate and push patches. 

So when creating the first set of patches, when gathering changes some code is triggered that does a push directly on the session (since it's already active and current) which leads to a second generation of patches is started, before the first one is finished. 

Not quite sure how to handle this. I have never considered collecting changes would trigger another push without scheduling it. So will need some more time to come up with some pattern or logic for this.

**Comment by careri** [comment]
So what you are saying is that when you're reading one of my properties I do a CalculatePatchAndPushOnWebSocket?

That might be so, I wrote this code before summer.

**Comment by chrhol** [comment]
> So what you are saying is that when you're reading one of my properties I do a CalculatePatchAndPushOnWebSocket?

Yes, that is what I think is happening. Maybe we could just check in Starcounter that we're already gathering changes and just skip further calls to push. The latest changes and patches will be sent anyway when it's finished.

**Comment by chrhol** [comment]
I've pushed a fix for this to develop and (hopefully) soon pnext. Should be available in the next nightly. 

@careri: It would be good if you could verify that it works as it should. I tried with HAD and as far as I can tell it seems to be working.

**Comment by chrhol** [comment]
Pushed to pnext now.

**Comment by careri** [comment]
I built your code, seems to be working nicely.

So just to figure out what I've done wrong. If I'm on a thread where Session.Current isn't null, I don't need to call CalculatePatchAndPushOnWebSocket?

If I'm doing work on a thread where Session.Current is null I need to call CalculatePatchAndPushOnWebSocket on all Sessions that should be updated with my work?

**Comment by careri** [comment]
This goes into the upcoming stable 2.2 right?

**Comment by careri** [comment]
FYI: I removed all calls CalculatePatchAndPushOnWebSocket when Session.Current != null.
Seems to work fine on 2.2 stable

**Comment by dan31** [comment]
@chrhol @warpech can you confirm it comes in ```2.2.1.3234```? If not, it shall come in ```2.2.2.x```.

**Comment by warpech** [comment]
This fix was added to `develop` 3 days after `2.2.1.3234`, so it will not be part of this stable release but the next one.

**Comment by chrhol** [comment]
Conclusion is that then the fix works, but is not included in 2.2.1. It will be included in 2.2.2 release. I consider this issue resolved then and will close it!
